src/server/worker/pipeline/common.go

```diff
@@ -0,0 +1,135 @@
+package pipeline
+
+import (
+	"archive/tar"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/pachyderm/pachyderm/src/client"
+	"github.com/pachyderm/pachyderm/src/client/pfs"
+	"github.com/pachyderm/pachyderm/src/client/pps"
+	"github.com/pachyderm/pachyderm/src/server/pkg/backoff"
+	"github.com/pachyderm/pachyderm/src/server/pkg/errutil"
+	"github.com/pachyderm/pachyderm/src/server/pkg/ppsconsts"
+	"github.com/pachyderm/pachyderm/src/server/worker/driver"
+	"github.com/pachyderm/pachyderm/src/server/worker/logs"
+)
+
+// RunUserCode will run the pipeline's user code until canceled by the context
+// - used for services and spouts. Unlike how the transform worker runs user
+// code, this does not set environment variables or collect stats.
+func RunUserCode(
+	ctx context.Context,
+	driver driver.Driver,
+	logger logs.TaggedLogger,
+) error {
+	return backoff.RetryUntilCancel(ctx, func() error {
+		// TODO: shouldn't this set up env like the worker does?
```

Comment:
This seems like it should be a yes, but I see that we do not do that currently (@jdoliner )

Reply:
I agree that this should setup env like the worker does... I'm a little unclear though is this not the worker? It's in the worker dir, is this specific to just the pipeline worker? I think ultimately this should replicate the env that we have in pre 2.0 releases.

Reply:
This is worker code, but it is shared for spouts and services - transform pipelines use a different code path - the goal was to replicate existing behavior, but I feel like the existing behavior might be wrong.